server/src/main/java/io/crate/planner/optimizer/collect/Optimizer.java

```diff
@@ -0,0 +1,156 @@
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.optimizer.collect;
+
+import io.crate.analyze.expressions.ExpressionAnalyzer;
+import io.crate.common.collections.Lists2;
+import io.crate.exceptions.ConversionException;
+import io.crate.expression.symbol.FunctionCopyVisitor;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.CoordinatorTxnCtx;
+import io.crate.metadata.Functions;
+import io.crate.planner.PlannerContext;
+import io.crate.planner.optimizer.collect.rule.MoveArrayLengthOnReferenceCastToLiteralCastInsideOperators;
+import io.crate.planner.optimizer.collect.rule.MoveReferenceCastToLiteralCastInsideOperators;
+import io.crate.planner.optimizer.collect.rule.MoveReferenceCastToLiteralCastOnAnyOperatorsWhenLeftIsReference;
+import io.crate.planner.optimizer.collect.rule.MoveReferenceCastToLiteralCastOnAnyOperatorsWhenRightIsReference;
+import io.crate.planner.optimizer.collect.rule.MoveSubscriptOnReferenceCastToLiteralCastInsideOperators;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Match;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.Version;
+
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class Optimizer {
+
+    public static Symbol optimizeQuery(PlannerContext plannerContext, Symbol query) {
+        Optimizer optimizer = new Optimizer(
+            plannerContext.functions(),
+            plannerContext.transactionContext(),
+            () -> Version.CURRENT,
+            List.of(
+                MoveReferenceCastToLiteralCastInsideOperators::new,
+                MoveReferenceCastToLiteralCastOnAnyOperatorsWhenRightIsReference::new,
+                MoveReferenceCastToLiteralCastOnAnyOperatorsWhenLeftIsReference::new,
+                MoveSubscriptOnReferenceCastToLiteralCastInsideOperators::new,
+                MoveArrayLengthOnReferenceCastToLiteralCastInsideOperators::new
+            )
+        );
+        return optimizer.optimize(query);
+    }
+
+    private static final Logger LOGGER = LogManager.getLogger(Optimizer.class);
+
+    private final List<Rule<?>> rules;
+    private final Supplier<Version> minNodeVersionInCluster;
+    private final Functions functions;
+    private final Visitor visitor = new Visitor();
+
+    public Optimizer(Functions functions,
+                     CoordinatorTxnCtx coordinatorTxnCtx,
+                     Supplier<Version> minNodeVersionInCluster,
+                     List<Function<BiFunction<String, List<Symbol>, Symbol>, Rule<?>>> rules) {
+        BiFunction<String, List<Symbol>, Symbol> functionResolver =
```

Comment:
Based on the types and name it is hard to guess what `String` and `List<Symbol>`, `Symbol` here are referring to. Maybe it would be worth to introduce a named specific interface?